src/crimson/admin/admin_socket.cc

```diff
@@ -0,0 +1,896 @@
+            // add 'failed' to the contents of out_buf? not what happens in the old code
+            return seastar::make_ready_future<bool>(false);
+          } else {
+            return std::move(fut);
+          }
+        }).then([&out, &out_buf, gatep](auto call_res) {
+
+          string outbuf_cont = out_buf.to_str();
+          uint32_t response_length = htonl(outbuf_cont.length());
+          logger().info("asok response length: {}", outbuf_cont.length());
+
+          return out.write((char*)&response_length, sizeof(uint32_t)).then([&out, outbuf_cont]() {
+
+            if (outbuf_cont.empty())
+              return out.write("    ");
+            else
+              return out.write(outbuf_cont.c_str());
+          });
+        });
+      });
+
+    } else {
+
+      // no command op-code identified
+      std::string no_opcode = "error: no command identified in incoming message ("s + cmdline + ")"s;
+      uint32_t response_length = htonl(no_opcode.length());
+      logger().info("asok f len: {}", no_opcode.length());
+
+      return out.write((char*)&response_length, sizeof(uint32_t)).
+        then([&out, no_opcode, response_length]() {
+          if (response_length)
+            return out.write(no_opcode.c_str());
+          else
+            return seastar::now();
+        });
+    }
+  });
+}
+
+seastar::future<> AdminSocket::handle_client(seastar::input_stream<char>& inp, seastar::output_stream<char>& out)
+{
+  //  RRR \todo safe read
+  //  RRR \todo handle old protocol (see original code) - is still needed?
+
+  return inp.read().
+    then( [&out, this](auto full_cmd) {
+
+      seastar::sstring cmd_line{full_cmd.begin(), full_cmd.end()};
+      logger().debug("{}: incoming asok string: {}\n", __func__, cmd_line);
+      return execute_line(cmd_line, out);
+
+    }).then([&out]() { return out.flush(); }).
+    finally([&out]() { return out.close(); }).
+    then([&inp,&out]() {
+      return inp.close();
+    }).handle_exception([](auto ep) {
+      logger().debug("exception on {}: {}", __func__, ep);
+      return seastar::make_ready_future<>();
+    }).discard_result();
+}
+
+seastar::future<> AdminSocket::init(const std::string& path)
+{
+#ifdef UNIT_TEST_OSD_ASOK
+  // in-line unit-testing:
+  std::ignore = seastar::async([this]() {
+     seastar::sleep(3s).wait();
+     asok_unit_testing::utest_run_1(this).wait();
+  });
+#endif
+
+  return seastar::async([this, path] {
+    auto serverfut = init_async(path);
+  });
+}
+
+
+seastar::future<> AdminSocket::init_async(const std::string& path)
+{
+  internal_hooks().get(); // we are executing in an async thread, thus OK to wait()
+
+  logger().debug("{}: asok socket path={}", __func__, path);
+  auto sock_path = seastar::socket_address{seastar::unix_domain_addr{path}};
+
+  return seastar::do_with(seastar::engine().listen(sock_path), [this](seastar::server_socket& lstn) {
+
+    return seastar::do_until([this](){ return do_die; }, [&lstn,this]() {
```

Comment:
This all looks like fully async IO?  Do we actually need seastar::async at all?

Reply:
Probably no. But it:
-  probably helpful when shutting the server down before OSD termination (I'll code this shortly and see),
-  seems cleaner, as this is indeed a separate role

Reply:
I'm not sure if this is still relevant, but we probably want to completely avoid seastar::thread unless absolutely necessary.  The performance implications don't actually matter here, but in general seastar::thread is less performant and we don't want to be designing things with that in mind.